### PR TITLE
app-misc/nnn: add 4.7

### PR DESCRIPTION
**Pull request by s-hamann**
This PR adds the latest upstream release of nnn.

Moreover, I added USE flags for some of the compile-time options that nnn offers.
I did not add USE flags that merely *reduce* functionality without removing dependencies (e.g. mouse or session support).
Nor did I add USE flags for the user patches that nnn offers. I think that would be a nice addition, but they did not seem to work in my testing and even broke the effects of other flags.
With the USE flags left at the default values, you'll get nnn with upstream's default configuration.

Finally, I made the ebuild install the plugins provided by upstream. Unfortunately, they do not work out-of-the-box, since (to the best of my knowledge) nnn will not pick up plugins from a global directory, but only from it's user-specific config directory.
I considered tying the plugins to a USE flag, but decided against it, since they are only about 300 KiB of script files.

### CODE REVIEW

File: app-misc/nnn/nnn-4.7.ebuild [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+SLOT="0"
+KEYWORDS="~amd64 ~ppc64 ~x86"
+IUSE="+readline pcre 8contexts icons nerdfonts emoji qsort"
+
+DEPEND="sys-libs/ncurses:=
+	pcre? ( dev-libs/libpcre )
+	readline? ( sys-libs/readline:= )
+	elibc_musl? ( sys-libs/fts-standalone )"
+RDEPEND="${DEPEND}"
+REQUIRED_USE="?? ( icons nerdfonts emoji )"
```
Comment (N-R-K) [inline app-misc/nnn/nnn-4.7.ebuild:22]: I think this should be `^^` as in only one of them should be set, because it's perfectly fine to build `nnn` without any icon support. See: https://devmanual.gentoo.org/ebuild-writing/variables/index.html#required_use

```suggestion
REQUIRED_USE="^^ ( icons nerdfonts emoji )"
```

File: app-misc/nnn/nnn-4.7.ebuild [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	insinto /usr/share/nnn
+	doins -r plugins
+}
+
+pkg_postinst() {
+	xdg_desktop_database_update
+
+	elog "nnn plugins are installed to /usr/share/nnn/plugins/, but nnn does not"
+	elog "load them fom this directory. You will need to copy/symlink them to"
+	elog "~/.config/nnn/plugins/ if you want to use them."
```
Comment (N-R-K) [inline app-misc/nnn/nnn-4.7.ebuild:67]: Some of the plugins also have their own dependencies (typically documented in the comment at the top of each plugins). Might be worth it to briefly mention that here as well.

**Review by N-R-K — CHANGES_REQUESTED** [review]

File: app-misc/nnn/nnn-4.7.ebuild (hunk 1 above)
Comment (s-hamann) [reply app-misc/nnn/nnn-4.7.ebuild:22]: Yes, that was my intention. The devmanual indicates that `^^` is "exactly one of" (XOR). `??` is not documented there, but it seems to enforce "at most one of", which is what we need here. Or am I wrong?

File: app-misc/nnn/nnn-4.7.ebuild (hunk 2 above)
Comment (s-hamann) [reply app-misc/nnn/nnn-4.7.ebuild:67]: Yes, that makes sense. Done.

File: app-misc/nnn/nnn-4.7.ebuild (hunk 1 above)
Comment (N-R-K) [reply app-misc/nnn/nnn-4.7.ebuild:22]: > Yes, that was my intention. The devmanual indicates that `^^` is "exactly one of" (XOR). `??` is not documented there, but it seems to enforce "at most one of", which is what we need here. Or am I wrong?

I can see `??` documentation just below.

image

Also now that I see the docs, I think I'm the one who's messing things up. `??` seems to be the right thing to use here. So ignore my previous comment :)

**Comment by scardracs** [comment]
I see this is a maintainer-needed package now. Do you mind maintaining it?

**Comment by scardracs** [comment]
> > I see this is a maintainer-needed package now. Do you mind maintaining it?
> 
> I think I can do that. I added myself to `metadata.xml` in a separate commit. Is that OK or would you prefer me to squash them into one?

It’s ok to maintain 2 different commits

File: app-misc/nnn/nnn-4.7.ebuild [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+SLOT="0"
+KEYWORDS="~amd64 ~ppc64 ~x86"
+IUSE="+readline pcre 8contexts icons nerdfonts emoji qsort"
+
+DEPEND="sys-libs/ncurses:=
+	pcre? ( dev-libs/libpcre )
+	readline? ( sys-libs/readline:= )
+	elibc_musl? ( sys-libs/fts-standalone )"
+RDEPEND="${DEPEND}"
+REQUIRED_USE="?? ( icons nerdfonts emoji )"
```
Comment (juippis) [inline app-misc/nnn/nnn-4.7.ebuild:22]: I'm not a huge fan of what's happening in here. Always try to avoid `REQUIRED_USE` whenever possible. In this case, users have to manually do steps outside setting use flags to get these to work anyway.

So my question is: does the build system prevent from these being toggled on? Since in this ebuild they don't pull any extra deps, can they be enabled conditionally and then put a `pkg_postinst` message saying enable one by symlinking it or something?

File: app-misc/nnn/nnn-4.7.ebuild [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+src_compile() {
+	emake \
+		O_NORL=$(usex readline 0 1) \
+		O_PCRE=$(usex pcre 1 0) \
+		O_CTX8=$(usex 8contexts 1 0) \
+		O_ICONS=$(usex icons 1 0) \
+		O_NERD=$(usex nerdfonts 1 0) \
+		O_EMOJI=$(usex emoji 1 0) \
+		O_QSORT=$(usex qsort 1 0) \
+		|| die "emake failed"
```
Comment (juippis) [inline app-misc/nnn/nnn-4.7.ebuild:40]:
```suggestion
	emake \
		O_NORL=$(usex readline 0 1) \
		O_PCRE=$(usex pcre 1 0) \
		O_CTX8=$(usex 8contexts 1 0) \
		O_ICONS=$(usex icons 1 0) \
		O_NERD=$(usex nerdfonts 1 0) \
		O_EMOJI=$(usex emoji 1 0) \
		O_QSORT=$(usex qsort 1 0) \
		|| die "emake failed"
```
`emake` does not need to `|| die` (since it handled that one on its own already), 
but an array form here would be much more cleaner.

File: app-misc/nnn/nnn-4.7.ebuild (hunk 3 above)
Comment (juippis) [inline app-misc/nnn/nnn-4.7.ebuild:22]: The Makefile has:
```
PKG_CONFIG ?= pkg-config
```
do BDEPEND on *virtual/pkgconfig*.

Comment (N-R-K) [reply app-misc/nnn/nnn-4.7.ebuild:22]: It's a build time toggle, yes. And the options are mutually exclusive with each other.

Comment (s-hamann) [reply app-misc/nnn/nnn-4.7.ebuild:23]: The build system will not allow enabling more than one of these. As far as I know, the respective icons are enabled unconditionally when nnn is built with support for them. They may not be displayed *correctly* (depending on the font setup), but *something* will show. Therefore, we need to option disable any single or none these flags.
If there is anything better suited than `REQUIRED_USE`, I'll happily switch to that. At the moment, I can't think of an alternative.

File: app-misc/nnn/nnn-4.7.ebuild (hunk 4 above)
Comment (s-hamann) [reply app-misc/nnn/nnn-4.7.ebuild]: I'm not sure if I understood you correctly. Please take another look.

File: app-misc/nnn/nnn-4.7.ebuild (hunk 3 above)
Comment (s-hamann) [reply app-misc/nnn/nnn-4.7.ebuild]: Done. Should I backport that to the existing ebuilds for nnn as well?

**Comment by N-R-K** [comment]
> Nor did I add USE flags for the user patches that nnn offers. I think that would be a nice addition, but they did not seem to work in my testing and even broke the effects of other flags.

Forgot to ask this (and it's a bit off topic), but can you specify what exactly was the problem? I can probably report it upstream and/or get it fixed if you provide some more details. @s-hamann

File: app-misc/nnn/nnn-4.7.ebuild (hunk 3 above)
Comment (juippis) [reply app-misc/nnn/nnn-4.7.ebuild:23]: You can "prioritize" the choice based on use flags selection - most likely just how upstream does it. But in this case since there's the additional `pkg_postinst` steps users have to do, I believe it'd be a messier user experience.

File: app-misc/nnn/nnn-4.7.ebuild [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+	elibc_musl? ( sys-libs/fts-standalone )"
+BDEPEND="virtual/pkgconfig"
+RDEPEND="${DEPEND}"
+REQUIRED_USE="?? ( icons nerdfonts emoji )"
+
+src_prepare() {
+	default
+	tc-export CC
+	use elibc_musl && append-flags "-lfts"
+	sed -i -e '/install: all/install:/' Makefile || die "sed failed"
```
Comment (juippis) [inline app-misc/nnn/nnn-4.7.ebuild:29]: Can you please add a comment above the sed why it's necessary? 
(also I see pretty much every distro has this sed, why can't this be fixed upstream? ping @N-R-K )

File: app-misc/nnn/nnn-4.7.ebuild [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+	if use icons; then
+		einfo "In order for file icons to work, your terminal needs to use icons-in-terminal."
+		einfo "See https://github.com/sebastiencs/icons-in-terminal"
+	elif use nerdfonts; then
+		einfo "In order for file icons to work, your terminal needs to use a patched nerdfont."
+		einfo "See https://www.nerdfonts.com/"
+	elif use emoji; then
+		einfo "In order for file icons to work, your terminal needs to use a font that"
+		einfo "includes standard unicode emoji."
+	fi
```
Comment (juippis) [inline app-misc/nnn/nnn-4.7.ebuild:82]: I wonder if we're better switching to `elog` here. IIRC `einfo` doesn't show if you specify `--jobs` for your emerge.

File: app-misc/nnn/nnn-4.7.ebuild (hunk 5 above)
Comment (s-hamann) [reply app-misc/nnn/nnn-4.7.ebuild:29]: To be honest, I have no idea why that line is necessary. It was apparently added by @Zlogene in 950b512dde3a2fef38e36329d2e8e7b70b8e2bb7 (nnn 3.6).
It replaced the command `sed -i -e '/strip/d' Makefile`. I see why this needed replacing: The `Makefile` changed to include 'strip' in a *lot* of lines. However, I don't see how the `all` target would strip anything, neither in 3.6 nor in 4.7.

The line does not seem to make a difference when building the package, so I removed it.

**Comment by s-hamann** [comment]
> > Nor did I add USE flags for the user patches that nnn offers. I think that would be a nice addition, but they did not seem to work in my testing and even broke the effects of other flags.
> 
> Forgot to ask this (and it's a bit off topic), but can you specify what exactly was the problem? I can probably report it upstream and/or get it fixed if you provide some more details. @s-hamann

Very good question. I just gave it a quick try and did not look further when it didn't work as expected, so I can't give you any details right now. I'll dig into this and report it upstream if it looks like an upstream bug.

File: app-misc/nnn/nnn-4.7.ebuild (hunk 5 above)
Comment (N-R-K) [reply app-misc/nnn/nnn-4.7.ebuild]: From a quick look, `all` is a prerequisite of `install` so that if someone runs `make install` it without compiling first, the makefile will do the compilation as well.

But I'm not sure what problem it's causing for distro packagers, and I cannot figure anything out from a quick look either, so I'm not sure what to fix here. /shrug

Comment (s-hamann) [reply app-misc/nnn/nnn-4.7.ebuild]: Update: That `sed` line did not do anything (missing `s`). However, it *is* necessary in combination with nnn's bundled patches. The `Makefile` applies the patches, compiles and then reverts the patches. That makes the source files newer than the compiled binary. The 'install' target therefore compiles nnn *again*. Since the ebuild does not pass the respective flags `src_install`, that second build was "vanilla", i.e. without any customization at all. That explains why the patches seemed to break the other USE flags.

TL;DR: I added the line back, fixed it and wrote a comment. I also added USE flags for nnn's bundled patches.

Comment (N-R-K) [reply app-misc/nnn/nnn-4.7.ebuild]: > Edit: See [jarun/nnn#1493](https://github.com/jarun/nnn/issues/1493).

Ah, I see. I forgot about the patching shenanigans - makes sense now. Though it's still a bit weird why the ebuild was doing it even though it didn't allow patches.

File: app-misc/nnn/nnn-4.7.ebuild [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+REQUIRED_USE="?? ( icons nerdfonts emoji )"
+
+src_prepare() {
+	default
+	tc-export CC
+	use elibc_musl && append-flags "-lfts"
+	# When one of nnn's bundled patches is used, the Makefile patches the
+	# source, compiles and then reverts the patch. This causes the source files
+	# to be newer than the compiled binary so that it get's compiled again for
+	# the 'install' target. Removing the dependency on 'all' fixes this.
```
Comment (N-R-K) [inline app-misc/nnn/nnn-4.7.ebuild:32]: Typically, the convention is to refer to the bug report instead.

```suggestion
	# See bug: https://github.com/jarun/nnn/issues/1493
```

Comment (s-hamann) [reply app-misc/nnn/nnn-4.7.ebuild]: Of course. I wrote that before I found the issue.

File: app-misc/nnn/nnn-4.7.ebuild (hunk 5 above)
Comment (juippis) [reply app-misc/nnn/nnn-4.7.ebuild]: Remember we allow user patching via */etc/portage/patches*, I imagine this will override them as well.

**Review by juippis — APPROVED** [review]
Thanks for your patience!
